src/LowerParallelTasks.cpp

```diff
@@ -0,0 +1,514 @@
+                                                {closure_arg, struct_type, make_const(UInt(32), struct_index)},
+                                                Call::PureIntrinsic),
+                                     wrapped_body);
+        struct_index++;
+    }
+    for (const auto &b : closure.buffers) {
+        type_args[struct_index + 2] = make_zero(type_of<void *>());
+        type_args[struct_index + 3] = make_zero(type_of<halide_buffer_t *>());
+        wrapped_body = LetStmt::make(b.first,
+                                     Call::make(type_of<void *>(), Call::load_struct_member,
+                                                {closure_arg, struct_type, make_const(UInt(32), struct_index)},
+                                                Call::PureIntrinsic),
+                                     wrapped_body);
+        wrapped_body = LetStmt::make(b.first + ".buffer",
+                                     Call::make(type_of<halide_buffer_t *>(), Call::load_struct_member,
+                                                {closure_arg, struct_type, make_const(UInt(32), struct_index + 1)},
+                                                Call::PureIntrinsic),
+                                     wrapped_body);
+        struct_index += 2;
+    }
+
+    Expr struct_type_decl = Call::make(Handle(), Call::declare_struct_type, type_args, Call::PureIntrinsic);
+    wrapped_body = Block::make(Evaluate::make(closure_arg), wrapped_body);
+    wrapped_body = LetStmt::make(closure_type_name, struct_type_decl, wrapped_body);
+
+    // TODO(zvookin): Figure out how we want to handle name mangling of closures.
+    // For now, the C++ backend makes them extern "C" so they have to be NameMangling::C.
+    LoweredFunc result{name, args, wrapped_body, LinkageType::External, NameMangling::C};
+    if (t.has_feature(Target::Debug)) {
+        debug_arguments(&result, t);
+    }
+    return result;
+}
+
+Expr allocate_closure(const std::string &name, const Closure &closure) {
+    std::vector<Expr> closure_elements;
+    std::vector<Expr> closure_types;
+
+    std::string buffer_t_type_name = unique_name("buffer_t_type");
+    Expr buffer_t_type = Variable::make(Handle(), buffer_t_type_name);
+
+    // Place holder for closure struct type
+    closure_elements.emplace_back(Expr());
+    closure_elements.emplace_back(1);
+    closure_types.emplace_back(unique_name(name + "_type"));
+    closure_types.emplace_back(1);
+
+    for (const auto &v : closure.vars) {
+        Expr var = Variable::make(v.second, v.first);
+        closure_elements.emplace_back(var);
+        closure_types.emplace_back(var);
+    }
+    for (const auto &b : closure.buffers) {
+        Expr ptr_var = Variable::make(type_of<void *>(), b.first);
+        closure_elements.emplace_back(ptr_var);
+        closure_elements.emplace_back(Call::make(type_of<halide_buffer_t *>(), Call::get_pointer_symbol_or_null,
```

Comment:
Why isn't this `Variable::make(type_of<halide_buffer_t *>(), b.first + ".buffer")`?

Why might the buffer exist sometimes and not exist other times, and it's OK if it doesn't exist?

If the buffer might not exist and that's OK for some reason, how about keeping a `Scope<Expr>` of all containing letstmts and using `scope.contains(buffer_name) ? scope.get(buffer_name) : reinterpret(Handle(), 0)`